carapace-server/src/main/java/org/carapaceproxy/server/Listeners.java

```diff
@@ -282,33 +275,59 @@ public void stop() {
         }
     }
 
-    private SslContext resolveSslContext(NetworkListenerConfiguration listener, String sniHostname) throws ConfigurationNotValidException {
-        int port = listener.getPort() + parent.getListenersOffsetPort();
-        String key = listener.getHost() + ":" + port + "+" + sniHostname;
-        if (LOG.isLoggable(Level.FINER)) {
-            LOG.log(Level.FINER, "resolve SNI mapping " + sniHostname + ", key: " + key);
+    private final class ListenerSniMapping implements io.netty.util.AsyncMapping<String, SslContext> {
+
+        private final Map<String, SslContext> listenerSslContexts = new ConcurrentHashMap<>();
+        private final NetworkListenerConfiguration listener;
+        private final int port;
+
+        public ListenerSniMapping(NetworkListenerConfiguration listener, int port) {
+            this.listener = listener;
+            this.port = port;
         }
-        try {
-            return sslContexts.computeIfAbsent(key, (k) -> {
+
+        @Override
+        public Future<SslContext> map(String sniHostname, Promise<SslContext> promise) {
+            try {
+                String key = listener.getHost() + ":" + port + "+" + sniHostname;
+                if (LOG.isLoggable(Level.FINER)) {
+                    LOG.log(Level.FINER, "resolve SNI mapping " + sniHostname + ", key: " + key);
+                }
                 try {
+                    SslContext sslContext = listenerSslContexts.get(key);
+                    if (sslContext != null) {
+                        return promise.setSuccess(sslContext);
+                    }
+
+                    sslContext = sslContexts.get(key);
+                    if (sslContext != null) {
+                        listenerSslContexts.put(key, sslContext);
+                        return promise.setSuccess(sslContext);
+                    }
+
                     SSLCertificateConfiguration choosen = chooseCertificate(sniHostname, listener.getDefaultCertificate());
                     if (choosen == null) {
                         throw new ConfigurationNotValidException("cannot find a certificate for snihostname " + sniHostname
                                 + ", with default cert for listener as '" + listener.getDefaultCertificate()
                                 + "', available " + currentConfiguration.getCertificates().keySet());
                     }
-                    return bootSslContext(listener, choosen);
-                } catch (ConfigurationNotValidException err) {
-                    throw new RuntimeException(err);
+                    sslContext = bootSslContext(listener, choosen);
```

Comment:
idk if you did it on purpose, but now there is a possibility where more than one listener will create the SslContext for the same key
btw this is a less locking-situation so I hope this will decrease latency (maybe with more CPU usages in some cases)